package/MDAnalysis/lib/pkdtree.py

```diff
@@ -145,75 +146,87 @@ def set_coords(self, coords):
-
-    def search(self, center, radius):
-        """Search all points within radius of center and its periodic images.
+        images = list()
+
+        if center_points.shape == (3,):
+            center_points = center_points.reshape((1,3))
+
+        # To-Do: parallel
+        for center_point in center_points:
+            # displacements are vectors we add to center_point to generate
+            # images "up" or "down" the central cell along the axis that we
+            # happen to be looking.
+
+            # distances to cell boundary planes passing through the origin
+            distances = np.dot(self._rm, center_point)
+            displacements = list(self._dm[np.where(distances < radius)[0]])
+            # distances to the remaining three cell boundary planes
+            distances = np.einsum('ij,ij->i', self._rm, self._dm - center_point)
+            displacements.extend(list(-self._dm[np.where(distances < radius)[0]]))
+            # If we have displacements along more than one axis, we have
+            # to combine them. This happens when center_point is close
+            # to any edge or vertex of the central cell.
+            # face case: n_displacements==1; no combination
+            # edge case: n_displacements==2; one extra displacement
+            # vertex case: n_displacements==3; five extra displacements
+            n_displacements = len(displacements)
+            if n_displacements > 1:
+                for start in range(n_displacements - 1, -1, -1):
+                    for i in range(start+1, len(displacements)):
+                        displacements.append(displacements[start]+displacements[i])
+            images.extend([center_point + d for d in displacements])
+        return images
+
+    def search(self, centers, radius):
+        """Search all points within radius of centers and its periodic images.
         Wrapping of center coordinates is enforced to enable comparison to
         wrapped coordinates of points in the tree.
 
         Parameter
         ---------
-        center: NumPy.array
-          origin around which to search for neighbors
+        centers: NumPy.array shape=(N,3)
+          origins around which to search for neighbors
         radius: float
           maximum distance around which to search for neighbors. The search
           radius is half the smallest periodicity if radius exceeds this value
         """
         if not self.built:
             raise RuntimeError('Unbuilt tree. Run tree.set_coords first')
-        if center.shape != (self.dim,):
-            raise ValueError('Expected a ({},) NumPy array'.format(self.dim))
+        if centers.shape == (self.dim, ):
+            centers = centers.reshape((1, self.dim))
+        wrapped_centers = apply_PBC(centers, self.box)
         self._indices = set()  # clear previous search
-        for c in self.find_centers(center, radius):
+        # To-Do: parallel
```

Comment:
I assume here you do spend some time on each iteration. Before you try to optimize this further I would suggest though to get exact numbers of the performance with respect to number of atoms and box size.